### DISCUSSION TITLE
How to declare record functions on a nested record

### DISCUSSION THREAD

**Discussion by Andre-LA** [teal-language/tl#648] [Q&A] [answered]
In the tutorial, there's this paragraph and example:

> You can also declare record functions after the record definition using the regular Lua colon or dot syntax, as long as you do it in the same scope block where the record type is defined

```lua
function Point:move(dx: number, dy: number)
   self.x = self.x + dx
   self.y = self.y + dy
end
```

This works, however, I'm curious if it's possible to do the same thing on a nested record like this?

```lua
local record Math
   record Point
      x: number
      y: number
   end
end

function Math.Point:move(dx: number, dy: number)
   self.x = self.x + dx
   self.y = self.y + dy
end
```

The code above gives the error "cannot add undeclared function 'move' outside of the scope where 'Math.Point' was originally declared", but I don't didn't understood how to declare it on the correct scope.

**Comment by Andre-LA** [comment]
Just after making the post I noticed how to do it, sorry 😅.

This solves the problem:

```lua
local record Math
   record Point
      x: number
      y: number
   end
end

local Point = Math.Point

function Point.new(x: number, y: number): Point
   local p: Point = {x=x, y=y}
   return setmetatable(p, {__index=Point})
end

function Point:move(dx: number, dy: number)
   self.x = self.x + dx
   self.y = self.y + dy
end

local p = Point.new(3,3)
p:move(2,3)
print(p.x,p.y) -- output: 5	6
```

I'll keep the thread open in case there's a more appropriate way.

**Comment by Andre-LA** [comment, marked as answer]
Making a correction here, the above method (by aliasing) doesn't works with nested records (with more depth than 1) and don't seems to be the apropriate way of declaring functions either.

The following way, described by the tutorial, it's reliable and works in any nesting level:

> If you want to define the function in a later scope (for example, if it is a callback to be defined by users of a module you are creating), you can declare the type of the function field in the record and fill it later from anywhere:
> 
> ```lua
> local record Obj
>    location: Point
>    draw: function(Obj)
> end
> ```

Closing.

**Comment by hishamhm** [comment]
This was a bug which is now fixed! Your original example now typechecks in `master`!

**Reply by Andre-LA** [reply]
Oh thanks! I didn't realised it was a bug 😅
